Strict object schema with a boolean property expects alternative, not boolean

<!--
  ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️
  You must complete this entire issue template to receive support. You MUST NOT remove, change, or replace the template with your own format. A missing or incomplete report will cause your issue to be closed without comment. Please respect the time and experience that went into this template. It is here for a reason. Thank you!
  ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️
-->

#### Support plan

<!--
We are here to help!

Community based support is, by its nature, limited to available community members able to help. Most community support issues are resolved within 2 weeks. Before submitting an issue, please review the various support options available at https://hapi.dev/support/. That page includes useful information about different ways to ask questions.
-->

* *is this issue currently blocking your project?* (yes/no): no
* *is this issue affecting a production system?* (yes/no): no

#### Context

* *node version*: 14.19.3
* *module version with issue*: 17.6.1
* *last module version without issue*: 17.6.0
* *environment* (e.g. node, browser, native): Node
* *used with* (e.g. hapi application, another framework, standalone, ...): Hapi
* *any other relevant information*: TypeScript 4.8.2

#### What are you trying to achieve or the steps to reproduce?

<!--
Describe your issue in detail, including full steps to reproduce the issue, any configuration, schemas, code samples, or inputs needed. Make sure to wrap all code examples in backticks so that they display correctly. Before submitting an issue, make sure to click on the Preview tab above to verify everything is formatted correctly.
-->

Joi 17.6.1's addition of `AlternativesSchema` for a strict object map has caused TypeScript to start reporting an error for boolean properties.

```ts
interface Test {
  enabled: boolean;
}

Joi.object<Test, true>({
  enabled: Joi.boolean(),
});
```

#### What was the result you got?

```
error TS2739: Type 'BooleanSchema' is missing the following properties from type 'AlternativesSchema': conditional, match, try
```

#### What result did you expect?

Successful compilation.

The problem appears to be that `IsUnion` returns true for booleans. I think that this may be because TypeScript treats `boolean` as the union of `true` and `false` (see the mentions of `'boolean'` within [checker.ts](https://github.com/microsoft/TypeScript/blob/main/src/compiler/checker.ts)), but I'm really not sure. See also [Stack Overflow's discussion](https://stackoverflow.com/a/73131776/25507) on the limitations of an `IsUnion` implementation.

To fix it, maybe reorder ObjectPropertiesSchema to check boolean before union?

[TypeScript playground link](https://www.typescriptlang.org/play?ts=4.8.2#code/PTAEAsBdIBwZwFwjpAhgYwNYHsBuBTAJwDMAbbAdwDp1sBbYVYAVgE4A2ADgHZu3gATM2YAGbgChIATxj5QASTgBVAHYBLbCoA8AFQA0oJaHwAPSPhUATOKB2gAvLYB8D8aHce7p81ZsBXFUwVShVQAH5QAG0lAF1jMwtrKJ04iOJUUjg5BFBIQj9s0HTM-ABucXFaFRRQVABGHMVVDW0VPzoAIyIXR2Ks8qqa1AFG5XVNLQ7sbFJ8VBUeooz+ipAIaHgkYBQMHAIScmpaBiZmAGZWYQAWM7rBYTFJGTkm8ZUBXQMjb0SbO0cdD03B4ABReBK+WoqKThYHuEHfCFJOxpZaFPIFACUcI87hyKnw+0x8R8ST6clRJVAOQxZQqg0goA6DQUYxaHzanW6DiWJQGmhqHRGrOamg+UxmcwWPPJ5XEaygsEQyDQWDwRDIlBo9EYLAu11u91EEmkskMbx02HkKnMhCy6EgLS0SkWwIRJN+UJhERBmBySkx9icuGwaks1NABKJHshIN9OTUKmIRAUgeDocsxIi8gjUaIwNNLzZmjOukWkRSMaS0QtVptRHtjomgNSvKyEdp9IFjPQLNeLVLnK6hEWssq3dA6GF-ZLk2ms3mo7RcrWAAVUIRGdhiOslVsdmr9pqjjrThweHxWEbHoWRW915vPoYq38eYCedE4j9IRXW+SO-kdIMqAlh9sWKgPpAWhDtyvTLuO1SMpY07gZBc6SouMrwUAA) for experimenting with IsUnion implementations

(I'm not sure about ObjectPropertiesSchema's union check for another reason: It thinks that `interface Test { value: 'a' | 'b' }` should use an AlternativesSchema, but it seems like a more idiomatic Joi implementation would be `Joi.string().valid('a', 'b')`. That's a separate issue and not currently affecting me, but I wanted to mention it in case it affects how this issue should be addressed. Thanks!)

--- a/lib/index.d.ts
+++ b/lib/index.d.ts
@@ -740,11 +740,30 @@ declare namespace Joi {
 
     type NullableType<T> = undefined | null | T
 
+    type IsPrimitiveSubset<T> =
+        [T] extends [string]
+        ? true
+        : [T] extends [number]
+        ? true
+        : [T] extends [bigint]
+        ? true
+        : [T] extends [boolean]
+        ? true
+        : [T] extends [symbol]
+        ? true
+        : [T] extends [null]
+        ? true
+        : [T] extends [undefined]
+        ? true
+        : false;
+
     type IsUnion<T, U extends T = T> =
       T extends unknown ? [U] extends [T] ? false : true : false;
 
+    type IsNonPrimitiveSubsetUnion<T> = true extends IsUnion<T> ? true extends IsPrimitiveSubset<T> ? false : true : false;
+
     type ObjectPropertiesSchema<T = any> =
-        true extends IsUnion<Exclude<T, undefined | null>>
+        true extends IsNonPrimitiveSubsetUnion<Exclude<T, undefined | null>>
         ? Joi.AlternativesSchema
         : T extends NullableType<string>
         ? Joi.StringSchema

--- a/test/index.ts
+++ b/test/index.ts
@@ -1263,11 +1263,13 @@ const userSchema2 = Joi.object<User>().keys({
 interface Comment {
     text: string;
     user: User;
+    isNew: boolean;
 }
 
 const commentSchemaObject = Joi.object<Comment, true>({
     text: Joi.string().required(),
-    user: userSchemaObject
+    user: userSchemaObject,
+    isNew: Joi.boolean().required(),
 });
 
 interface Comment2 {
@@ -1287,11 +1289,15 @@ const commentSchemaObject2 = Joi.object<Comment2, true>({
 interface CommentWithAlternatives {
     text: string;
     user: string | User;
+    type: 'topLevel' | 'pingback' | 'reply';
+    reported: boolean | number;
 }
 
 const commentWithAlternativesSchemaObject = Joi.object<CommentWithAlternatives, true>({
     text: Joi.string().required(),
     user: Joi.alternatives(Joi.string(), userSchemaObject),
+    type: Joi.string().required().valid('topLevel', 'pingback', 'reply'),
+    reported: Joi.alternatives(Joi.boolean(), Joi.number()),
 });
 
 expect.error(userSchema2.keys({ height: Joi.number() }));

